service/s3/unmarshal.go

```diff
@@ -0,0 +1,114 @@
+package s3
+
+import (
+	"encoding/xml"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/awserr"
+	"github.com/aws/aws-sdk-go-v2/internal/sdkio"
+	restlegacy "github.com/aws/aws-sdk-go-v2/private/protocol/rest"
+)
+
+/*
+   GetObject operation doesn't support 200 status code with errors.
+
+*/
+
+// protoGetObjectUnmarshaler defines unmarshaler for ProtoGetObject operation
+type protoGetObjectUnmarshaler struct {
+	output *GetObjectOutput
+}
+
+// unmarshalOperation is the top level method used with a handler stack to unmarshal an operation response
+// This method calls appropriate unmarshal shape functions as per the output shape and protocol used by the service.
+func (u protoGetObjectUnmarshaler) unmarshalOperation(r *aws.Request) {
+	defer r.HTTPResponse.Body.Close()
+	// isRequestError checks if operation response returned a error response
+	if isRequestError(r) {
+		// if startToken is nil, it would mean there is no xml response body
+		if r.HTTPResponse.Body == nil {
+			// fall back to status code error message
+			r.Error = getStatusCodeErrorMessage(r)
+			return
+		}
+
+		var buff []byte
+		ringBuff := sdkio.NewRingBuffer(buff)
+		body := io.TeeReader(r.HTTPResponse.Body, ringBuff)
+		decoder := xml.NewDecoder(body)
+		startToken, err := decoder.Token()
+		if err != nil {
+			r.Error = awserr.New(aws.ErrCodeSerialization, "Failed to decode response body with invalid XML", err)
+			return
+		}
+
+		r.Error = unmarshalErrorPrototype(r, decoder, startToken)
```

Comment:
Serialization errors encountered here, will not include the ring buffer, but should.